lib/puppet/resource_api/transport.rb

```diff
@@ -37,17 +36,32 @@ def connect(name, connection_info)
   module_function :connect # rubocop:disable Style/AccessModifierDeclarations
 
   def self.validate(name, connection_info)
-    @transports ||= {}
+    init_transports
     require "puppet/transport/schema/#{name}" unless @transports.key? name
-    transport_schema = @transports[name]
-    raise Puppet::DevError, 'Transport for `%{target}` not registered' % { target: name } if transport_schema.nil?
+    transport_schema = @transports[@environment][name]
+    raise Puppet::DevError, 'Transport for `%{target}` not registered with `%{environment}`' %  {
+                                target: name,
+                                environment: @environment
+                              } if transport_schema.nil?
 
     transport_schema.check_schema(connection_info)
     transport_schema.validate(connection_info)
   end
 
   def self.get_context(name)
     require 'puppet/resource_api/puppet_context'
-    Puppet::ResourceApi::PuppetContext.new(@transports[name])
+    Puppet::ResourceApi::PuppetContext.new(@transports[@environment][name])
+  end
+
+  def init_transports
+    lookup = Puppet.lookup(:current_environment) if Puppet.respond_to? :lookup
+    @environment =  if lookup.nil?
+                      :transports_default
+                    else
+                      lookup.name
+                    end
+    @transports ||= {}
+    @transports[@environment] ||= {}
   end
+  module_function :init_transports # rubocop:disable Style/AccessModifierDeclarations
```

Comment:
As this is not part of the official API, this should be marked `private`, not `module_function`. Now that I think about this, the `get_context` and `validate` above, too.